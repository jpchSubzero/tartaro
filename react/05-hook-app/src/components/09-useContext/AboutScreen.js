import React, { useContext } from 'react';
import { UserContext } from './UserContext';


export const AboutScreen = () => {

    // 1. Como dejamos de enviar solo el valor y ahora es el state, destructuramos el objeto para tener user y setUser. En este caso solo user porque no se utiliza setUser
    const { user, setUser } = useContext(UserContext);

    const handleClick = () => {
        setUser({});
    };

    return (
        <div>
            <h1>AboutScreen</h1>
            <hr />

            {/* 3. Agregamos un bot√≥n para borrar el user */}
            <button className="btn btn-danger" onClick={ handleClick }>Log Out</button>

            {/* 2. Agregar el user como json */}
            <pre className="container">{ JSON.stringify(user, null, 3) }</pre>

        </div>    
    )
}
