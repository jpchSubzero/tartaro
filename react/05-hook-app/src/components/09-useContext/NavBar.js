import React from 'react';
// import { Link } from "react-router-dom";

// Quitar el Link y agregar el NavLink que sirve para detectar si esta activo o no y poder utilizar clases de estilo
import { NavLink } from "react-router-dom";

export const NavBar = () => {
    return (
        // Definir el elemento para navegar nav y una lista de links
        // <nav>
        //     <ul>
        //         {/* Esto mostrará una lista sencilla donde estará el navbar */}
        //         {/* <li>Home</li>
        //         <li>Login</li>
        //         <li>About</li> */}

        //         {/* Esta forma funciona pero basado en petición al servidor, es decir, refresca la página */}
        //         {/* <li><a href="/">Home</a></li>
        //         <li><a href="/login">Login</a></li>
        //         <li><a href="/about">About</a></li> */}

        //         {/* Para evitar el refresh se utilzar Link del react-router-dom. Funciona como el a pero en lugar de a es Link y en lugar de href es to */}
        //         <li><Link to="/">Home</Link></li>
        //         <li><Link to="/login">Login</Link></li>
        //         <li><Link to="/about">About</Link></li>
        //     </ul>
        // </nav>

        // Una vez que funcionen los links, reemplazamos todo lo anterior por el navbar de bootstrap. Usar el dark y cambiar class por className ya que class es una palabra reservada de react
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
            <div className="container-fluid">
                {/* <Link className="navbar-brand" to="/">useContextApp</Link> */}

                {/* Utilizar NavLink y atributo activeClassName */}
                <NavLink activeClassName="active" className="navbar-brand" to="/">useContextApp</NavLink>
                <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarNavAltMarkup" aria-controls="navbarNavAltMarkup" aria-expanded="false" aria-label="Toggle navigation">
                    <span className="navbar-toggler-icon"></span>
                </button>                
                <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
                <div className="navbar-nav">
                    {/* <Link className="nav-link" aria-current="page" to="/">Home</Link>
                    <Link className="nav-link" aria-current="page" to="/login">Login</Link>
                    <Link className="nav-link" aria-current="page" to="/about">About</Link> */}

                    {/* Utilizar NavLink y atributo activeClassName. Utilizar exact para que vaya a la ruta explicitamente y no confunda / que tiene todas. Todo esto para el link activo se puede omitir con aria-current="page" de bootstrap pero es bueno saber que link esta activo */}
                    <NavLink exact activeClassName="active" className="nav-link" to="/">Home</NavLink>
                    <NavLink exact activeClassName="active" className="nav-link" to="/login">Login</NavLink>
                    <NavLink exact activeClassName="active" className="nav-link" to="/about">About</NavLink>
                </div>
                </div>
            </div>
        </nav>            
    )
}

// Tarea 1
// Después de copiar el navbar de bootstrap y usar el tema dark, hacer que funcionen los links