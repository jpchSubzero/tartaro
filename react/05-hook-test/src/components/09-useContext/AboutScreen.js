import React, { useContext } from 'react';
import { UserContext } from './UserContext';


export const AboutScreen = () => {
    const { user, setUser } = useContext(UserContext);

    const handleClick = () => {
        setUser({});
    };

    return (
        <div>
            <h1>AboutScreen</h1>
            <hr />

            <button className="btn btn-danger" onClick={ handleClick }>Log Out</button>

            <pre className="container">{ JSON.stringify(user, null, 3) }</pre>

        </div>    
    )
}
