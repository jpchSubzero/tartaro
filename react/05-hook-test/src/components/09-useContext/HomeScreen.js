import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const HomeScreen = () => {

    // 1. Agregarmos la referencia al contexto con el hook useContext. Devuelve el valor asignado en el componente que se defini√≥ para almacenar, en ese caso UserContext
    // const userContext = useContext(UserContext);
    
    // 2. Como dejamos de enviar solo el valor y ahora es el state, destructuramos el objeto para tener user y setUser. En este caso solo user porque no se utiliza setUser
    const { user } = useContext(UserContext);

    console.log(user);

    return (
        <div>
            <h1>HomeScreen</h1>
            <hr />

            {/* 3. Agregarmos un <pre> para ver los valores del user. La clase container es para que se vea mejor */}
            <pre className="container">{ JSON.stringify(user, null, 3) }</pre>
        </div>
    )
}
