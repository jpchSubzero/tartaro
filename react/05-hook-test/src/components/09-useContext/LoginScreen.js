import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const LoginScreen = () => {
    const { setUser } = useContext(UserContext);

    const newUser = {
        id: 1234,
        name: 'Nombre del usuario',
        email: 'usuario@usuario.com'
    };

    return (
        <div>
            <h1>LoginScreen</h1>
            <hr />

            <button className="btn btn-primary" onClick={() => setUser(newUser)}>Log In</button>
        </div>    
    )
}

