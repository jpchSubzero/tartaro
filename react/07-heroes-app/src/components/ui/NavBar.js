// Sacar el cÃ³digo para no tener que volver a hacer todo

import React from 'react';
import { Link, NavLink } from 'react-router-dom';

export const Navbar = () => {
    return (
        <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
            <div className="container-fluid">
                <Link className="navbar-brand" to="/">Asociaciones</Link>
                <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
                    <span className="navbar-toggler-icon"></span>
                </button>
                <div className="collapse navbar-collapse" id="navbarSupportedContent">
                    <div className="navbar-nav">
                        <NavLink 
                            activeClassName="active"
                            className="nav-item nav-link" 
                            exact
                            to="/marvel"
                        >Marvel
                        </NavLink>

                        <NavLink 
                            activeClassName="active"
                            className="nav-item nav-link" 
                            exact
                            to="/dc"
                        >DC
                        </NavLink>
                    </div>
                    <div className="d-flex flex-row-reverse w-100">
                        <ul className="navbar-nav">
                            <NavLink 
                                activeClassName="active"
                                className="nav-item nav-link" 
                                exact
                                to="/login"
                            >Logout
                            </NavLink>
                        </ul>
                    </div>
                </div>
            </div>
        </nav>
    )
}

