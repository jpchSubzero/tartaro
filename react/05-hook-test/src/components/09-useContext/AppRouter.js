import React from 'react';
import { AboutScreen } from './AboutScreen';
import { LoginScreen } from './LoginScreen';
import { HomeScreen } from './HomeScreen';

// Importar componentes para el uso de router. Se puede copiar directamente de la documentación.
// https://reactrouter.com/web/guides/quick-start
// as Router es para crear un alias y no utilizar todo el nombre del componente BrowserRouter.
// Para evitar los warnings de Link es mejor quitarlo porque se lo va a utilizar en el navbar
import { BrowserRouter as Router, Switch, Route, Redirect } from "react-router-dom";
import { NavBar } from './NavBar';

export const AppRouter = () => {
    return (
        // Especifica que es un router
        <Router>
            {/* Contenedor de todos los componentes que se vayan a utilizar. Es opcional */}
            <div>
                {/* 6. Agregar el NavBar después de probar las rutas manualmente */}
                <NavBar />

                {/* 7. Agregar un div container para que mejore la presentación */}
                <div className="container mt-5">
                    {/* Definir las rutas a utilizar */}
                    <Switch>
                        {/* 1. Es necesario el / en la ruta */}
                        {/* <Route path="/about" component={ AboutScreen } />
                        <Route path="/login" component={ LoginScreen } /> */}

                        {/* 2. Si lo ponemos primero siempre se va a mostrar el home porque valida por coincidencia parcial y si el path tiene / se iria al home. Para solucionar se puede ordenar los Route del más específico al más general (/user/create, /user/update, /about, /login, /user, /) o agregar el atributo exact a Route para que haga una evaluación completa de la ruta */}
                        {/* <Route path="/" component={ HomeScreen } /> */}

                        <Route exact path="/" component={ HomeScreen } />
                        <Route exact path="/about" component={ AboutScreen } />
                        <Route exact path="/login" component={ LoginScreen } />

                        {/* 3. Ruta por defecto para manejar rutas que no existen y deben llevar a una por defecto, por lo general a la página de Not Found. No cambia el path del navegador */}
                        {/* Comparando con cualquiera que no exista */}
                        {/* <Route exact path='*' component={HomeScreen} /> */}

                        {/* 4. Simplemente enviar a la ruta por defecto si no encuentra las anteriores */}
                        {/* <Route component={HomeScreen} /> */}

                        {/* 5. Esta opción redirecciona a una ruta por defecto. Implica refresh pero tambien cambia el path del navegador */}
                        <Redirect to="/" />
                    </Switch>
                </div>
            </div>
        </Router>
    )
}
