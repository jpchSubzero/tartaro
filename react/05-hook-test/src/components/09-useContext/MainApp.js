import React, { useState } from 'react';
import { AppRouter } from './AppRouter';
import { UserContext } from './UserContext';

export const MainApp = () => {

    // 3. Agregamos el objeto/información a compartir
    // const user = {
    //     id: 1234,
    //     name: 'Nombre del usuario',
    //     email: 'usuario@usuario.com'
    // }

    // 5. Reemplazamos el objeto por el state. COmo estabamos enviando user y el state tiene el mismo nombre app va a seguir igual solo que con un objeto vacío
    const [user, setUser] = useState({});

    return (
        // <div>
        //     <h1>MainApp</h1>
        //     <hr />
        //     <AppRouter />
        // </div>    

        // 1. Podemos dejar solo el AppRouter ya que es quien renderizará lo que se vaya pidiendo
        // <AppRouter />

        // 2. Utilizar el contexto creado para manejar la comunicación entre componentes que estan al mismo nivel. Al revisar en la consola de react se ve una estructura mucho más compleja que antes
        // <UserContext.Provider>
        //     <AppRouter />
        // </UserContext.Provider>

        // 4. Enviamos la información a través del atributo value. Al reviasr en consola se ve que ya tiene la prop el contexto
        // <UserContext.Provider value={ user }>

        // 6. Enviamos tanto el valor como la función en un objeto nuevo, { user, setUser }. Como usamos los mismo nombres se omite : { user:user, setUser:setUser }
        <UserContext.Provider value={ { user, setUser } }>
            <AppRouter />
        </UserContext.Provider>
    )
}
