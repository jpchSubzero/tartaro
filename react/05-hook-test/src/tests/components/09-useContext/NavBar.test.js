import React from 'react';
import { shallow } from 'enzyme';
import { NavBar } from '../../../components/09-useContext/NavBar';

describe('Pruebas en NavBar', () => {
    test('Debe presentar correctamente', () => {
        const wrapper = shallow(<NavBar />);
        expect(wrapper).toMatchSnapshot();
    });
});
