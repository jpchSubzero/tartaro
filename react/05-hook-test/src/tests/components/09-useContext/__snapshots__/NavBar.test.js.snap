// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en NavBar Debe presentar correctamente 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <div
    className="container-fluid"
  >
    <NavLink
      activeClassName="active"
      className="navbar-brand"
      to="/"
    >
      useContextApp
    </NavLink>
    <button
      aria-controls="navbarNavAltMarkup"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-bs-target="#navbarNavAltMarkup"
      data-bs-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarNavAltMarkup"
    >
      <div
        className="navbar-nav"
      >
        <NavLink
          activeClassName="active"
          className="nav-link"
          exact={true}
          to="/"
        >
          Home
        </NavLink>
        <NavLink
          activeClassName="active"
          className="nav-link"
          exact={true}
          to="/login"
        >
          Login
        </NavLink>
        <NavLink
          activeClassName="active"
          className="nav-link"
          exact={true}
          to="/about"
        >
          About
        </NavLink>
      </div>
    </div>
  </div>
</nav>
`;
